fragment/fragment/src/main/java/androidx/fragment/app/strictmode/FragmentStrictMode.java

```diff
@@ -0,0 +1,197 @@
+            }
+
+            /**
+             * Crash the whole process on violation. This penalty runs at the end of all enabled
+             * penalties so you'll still get to see logging or other violations before the process
+             * dies.
+             */
+            @NonNull
+            public Builder penaltyDeath() {
+                flags.add(Flag.PENALTY_DEATH);
+                return this;
+            }
+
+            /** Call #{@link OnViolationListener#onViolation} for every violation. */
+            @NonNull
+            public Builder penaltyListener(@NonNull OnViolationListener listener) {
+                this.listener = listener;
+                return this;
+            }
+
+            /**
+             * Construct the Policy instance.
+             *
+             * <p>Note: if no penalties are enabled before calling <code>build</code>, {@link
+             * #penaltyLog} is implicitly set.
+             */
+            @NonNull
+            public Policy build() {
+                if (listener == null && !flags.contains(Flag.PENALTY_DEATH)) {
+                    penaltyLog();
+                }
+                return new Policy(flags, listener);
+            }
+        }
+    }
+
+    /** Returns the current default policy. */
+     @NonNull
+    public static Policy getDefaultPolicy() {
+        return defaultPolicy;
+    }
+
+    /**
+     * Sets the policy for what actions should be detected, as well as the penalty if such actions
+     * occur.
+     *
+     * @param policy the policy to put into place
+     */
+    public static void setDefaultPolicy(@NonNull Policy policy) {
+        defaultPolicy = policy;
+    }
+
+    private static Policy getNearestPolicy(@Nullable Fragment fragment) {
+        while (fragment != null) {
+            if (fragment.isAdded()) {
+                FragmentManager fragmentManager = fragment.getParentFragmentManager();
+                if (fragmentManager.getStrictModePolicy() != null) {
+                    return fragmentManager.getStrictModePolicy();
+                }
+            }
+            fragment = fragment.getParentFragment();
+        }
+        return defaultPolicy;
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    static void onPolicyViolation(@NonNull Fragment fragment, @NonNull Violation violation) {
+        onPolicyViolation(getNearestPolicy(fragment), violation);
+    }
+
+    private static void onPolicyViolation(@NonNull Policy policy, @NonNull Violation violation) {
+        if (policy.flags.contains(Flag.PENALTY_LOG)) {
+            Log.d(TAG, "FragmentStrictMode policy violation: ", violation);
+        }
+
+        if (policy.listener != null) {
+            policy.listener.onViolation(violation);
+        }
+
+        if (policy.flags.contains(Flag.PENALTY_DEATH)) {
+            Log.e(TAG, "FragmentStrictMode policy violation with PENALTY_DEATH - shutting down.");
+            Process.killProcess(Process.myPid());
```

Comment:
Killing the process might be extreme here, I think we should throw the Violation instead.

Reply:
I got this from `StrictMode`, that's how it reacts to `penaltyDeath` violations: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/os/StrictMode.java;l=2308-2312?q=strictmode

My guess as to why it isn't an exception would be to bypass any `Thread.UncaughtExceptionHandler` that might be registered. But I'd also be fine with just throwing here.

Reply:
`Process.killProcess` means that any crash analytics won't pick up these exceptions, which I think is a big downside. Let's just throw.